pkg/gadgets/tcptop/tracer/tracer.go

```diff
@@ -0,0 +1,277 @@
+	}
+
+	consts := map[string]interface{}{
+		"filter_by_mnt_ns": filterByMntNs,
+		"target_pid":       t.config.TargetPid,
+		"target_family":    t.config.TargetFamily,
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil {
+		return fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	opts := ebpf.CollectionOptions{
+		Maps: ebpf.MapOptions{
+			PinPath: filepath.Dir(t.config.MountnsMap),
+		},
+	}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		return fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	t.tcpSendmsgLink, err = link.Kprobe("tcp_sendmsg", t.objs.TcpSendmsg)
+	if err != nil {
+		return fmt.Errorf("error opening kprobe: %w", err)
+	}
+
+	t.tcpCleanupRbufLink, err = link.Kprobe("tcp_cleanup_rbuf", t.objs.TcpCleanupRbuf)
+	if err != nil {
+		return fmt.Errorf("error opening kprobe: %w", err)
+	}
+
+	t.run()
+
+	return nil
+}
+
+func (t *Tracer) nextStats() ([]types.Stats, error) {
+	stats := []types.Stats{}
+
+	var prev *C.struct_ip_key_t = nil
+	key := C.struct_ip_key_t{}
+	ips := t.objs.IpMap
+
+	defer func() {
+		// delete elements
+		err := ips.NextKey(nil, unsafe.Pointer(&key))
+		if err != nil {
+			return
+		}
+
+		for {
+			if err := ips.Delete(key); err != nil {
+				return
+			}
+
+			prev = &key
+			if err := ips.NextKey(unsafe.Pointer(prev), unsafe.Pointer(&key)); err != nil {
+				return
+			}
+		}
+	}()
+
+	// gather elements
+	err := ips.NextKey(nil, unsafe.Pointer(&key))
+	if err != nil {
+		if errors.Is(err, ebpf.ErrKeyNotExist) {
+			return stats, nil
+		}
+		return nil, fmt.Errorf("error getting next key: %w", err)
+	}
+
+	for {
+		val := C.struct_traffic_t{}
+		if err := ips.Lookup(key, unsafe.Pointer(&val)); err != nil {
+			return nil, err
+		}
+
+		srcAddr := C.src_addr(&key)
+		defer C.free(unsafe.Pointer(srcAddr))
+		dstAddr := C.dst_addr(&key)
+		defer C.free(unsafe.Pointer(dstAddr))
+
+		stat := types.Stats{
+			Saddr:     C.GoString(srcAddr),
+			Daddr:     C.GoString(dstAddr),
+			MountNsID: uint64(key.mntnsid),
+			Pid:       int32(key.pid),
+			Comm:      C.GoString(&key.name[0]),
+			Sport:     uint16(key.lport),
+			Dport:     uint16(key.dport),
+			Family:    uint16(key.family),
+			Sent:      uint64(val.sent),
+			Received:  uint64(val.received),
+		}
+
+		container := t.resolver.LookupContainerByMntns(stat.MountNsID)
+		if container != nil {
+			stat.Container = container.Name
+			stat.Pod = container.Podname
+			stat.Namespace = container.Namespace
+			stat.Node = t.config.Node
+		}
+
+		stats = append(stats, stat)
+
+		prev = &key
+		if err := ips.NextKey(unsafe.Pointer(prev), unsafe.Pointer(&key)); err != nil {
+			if errors.Is(err, ebpf.ErrKeyNotExist) {
+				break
+			}
+			return nil, fmt.Errorf("error getting next key: %w", err)
+		}
+	}
+
+	types.SortStats(stats, t.config.SortBy)
+
+	return stats, nil
+}
+
+func (t *Tracer) run() {
+	ticker := time.NewTicker(t.config.Interval)
+
+	go func() {
+		for {
+			select {
+			case <-t.done:
```

Comment:
I suppose that we close the channel in `Trace.Stop()`, this event is triggered. However, when it happens, we are not breaking the `for` 🤔. I am not sure how this channel is being used here 😅.

cc @mauriciovasquezbernal

Reply:
ah, good catch! The filetop gadget had a break there and was removed in https://github.com/kinvolk/inspektor-gadget/commit/310bda79dd918b32941ccdaae4a6a3313d12dbf0#diff-ef952568eac284fe8563fe7c3ac80b06b40bc50440b1825f1e7046238ce0ea4cL229 because it was wrong and was breaking the switch statement instead of the for loop. I think we could just return there to finish the go rountine.

Reply:
Done!